gateway/fedimint-gateway-client/src/lib.rs

```diff
@@ -7,346 +10,317 @@ use fedimint_gateway_common::{
-    PaymentLogPayload, PaymentLogResponse, PaymentSummaryPayload, PaymentSummaryResponse,
-    RECEIVE_ECASH_ENDPOINT, ReceiveEcashPayload, ReceiveEcashResponse, SEND_ONCHAIN_ENDPOINT,
-    SET_FEES_ENDPOINT, SPEND_ECASH_ENDPOINT, STOP_ENDPOINT, SendOnchainRequest, SetFeesPayload,
-    SpendEcashPayload, SpendEcashResponse, WITHDRAW_ENDPOINT, WithdrawPayload, WithdrawResponse,
+    GetInvoiceRequest, GetInvoiceResponse, IrohGatewayRequest, IrohGatewayResponse,
+    LEAVE_FED_ENDPOINT, LIST_CHANNELS_ENDPOINT, LIST_TRANSACTIONS_ENDPOINT, LeaveFedPayload,
+    ListTransactionsPayload, ListTransactionsResponse, MNEMONIC_ENDPOINT, MnemonicResponse,
+    OPEN_CHANNEL_ENDPOINT, OpenChannelRequest, PAY_INVOICE_FOR_OPERATOR_ENDPOINT,
+    PAY_OFFER_FOR_OPERATOR_ENDPOINT, PAYMENT_LOG_ENDPOINT, PAYMENT_SUMMARY_ENDPOINT,
+    PayInvoiceForOperatorPayload, PayOfferPayload, PayOfferResponse, PaymentLogPayload,
+    PaymentLogResponse, PaymentSummaryPayload, PaymentSummaryResponse, RECEIVE_ECASH_ENDPOINT,
+    ReceiveEcashPayload, ReceiveEcashResponse, SEND_ONCHAIN_ENDPOINT, SET_FEES_ENDPOINT,
+    SPEND_ECASH_ENDPOINT, STOP_ENDPOINT, SendOnchainRequest, SetFeesPayload, SpendEcashPayload,
+    SpendEcashResponse, V1_API_ENDPOINT, WITHDRAW_ENDPOINT, WithdrawPayload, WithdrawResponse,
 };
+use iroh::Endpoint;
+use iroh::endpoint::Connection;
 use lightning_invoice::Bolt11Invoice;
 use reqwest::{Method, StatusCode};
 use serde::Serialize;
 use serde::de::DeserializeOwned;
 use thiserror::Error;
 
 pub struct GatewayRpcClient {
-    /// Base URL to gateway web server
-    /// This should include an applicable API version, e.g. http://localhost:8080/v1
     base_url: SafeUrl,
-    /// A request client
+    iroh_connector: Option<GatewayIrohConnector>,
     client: reqwest::Client,
-    /// Optional gateway password
     password: Option<String>,
 }
 
-impl GatewayRpcClient {
-    pub fn new(versioned_api: SafeUrl, password: Option<String>) -> Self {
-        Self {
-            base_url: versioned_api,
-            client: reqwest::Client::new(),
+#[derive(Debug, Clone)]
+struct GatewayIrohConnector {
+    node_id: iroh::NodeId,
+    endpoint: Endpoint,
+    password: Option<String>,
+}
+
+impl GatewayIrohConnector {
+    pub async fn new(iroh_pk: iroh::PublicKey, password: Option<String>) -> anyhow::Result<Self> {
+        let builder = Endpoint::builder().discovery_dht().discovery_n0();
+        let endpoint = builder.bind().await?;
+
+        Ok(Self {
+            node_id: iroh_pk,
+            endpoint,
             password,
-        }
+        })
     }
 
-    pub fn with_password(&self, password: Option<String>) -> Self {
-        GatewayRpcClient::new(self.base_url.clone(), password)
+    async fn connect(&self) -> anyhow::Result<Connection> {
+        let connection = self
+            .endpoint
+            .connect(self.node_id, FEDIMINT_GATEWAY_ALPN)
+            .await?;
+        // TODO: Spawn connection monitoring?
```

Comment:
This didn't seem worth it to add now, but might be necessary once we add `client -> gateway` over Iroh.